File: src/main/java/com/tambapps/http/restclient/AbstractRestClient.java
package com.tambapps.http.restclient;

import com.tambapps.http.restclient.request.RestRequest;
import com.tambapps.http.restclient.response.HttpHeaders;
import com.tambapps.http.restclient.response.handler.ResponseHandler;
import com.tambapps.http.restclient.response.RestResponse;
import com.tambapps.http.restclient.util.IOUtils;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Abstract class for a rest client
 */
public class AbstractRestClient {

  private final String baseUrl;
  private String jwt = null;

  /**
   *
   * @param baseUrl the base url of the rest api
   */
  protected AbstractRestClient(String baseUrl) {
    this.baseUrl = baseUrl;
  }

  private URL getUrl(String endpoint) throws MalformedURLException {
    if (baseUrl.endsWith("/")) {
      return new URL(
          endpoint.startsWith("/") ? baseUrl + endpoint.substring(1) : baseUrl + endpoint);
    } else {
      return new URL(baseUrl + (endpoint.startsWith("/") ?
          baseUrl + endpoint :
          baseUrl + '/' + endpoint));
    }
  }

  private HttpURLConnection prepareConnection(RestRequest request) throws IOException {
    HttpURLConnection connection =
        (HttpURLConnection) getUrl(request.getEndpoint()).openConnection();
    connection.setRequestMethod(request.getMethod());
    for (Map.Entry<String, String> header : request.getHeaders().entrySet()) {
      connection.setRequestProperty(header.getKey(), header.getValue());
    }
    if (jwt != null) {
      connection.setRequestProperty("Authorization", "Bearer " + jwt);
    }
    if (request.getTimeout() != null) {
      connection.setConnectTimeout(request.getTimeout());
    }
    return connection;
  }

  protected  <T> RestResponse<T> doExecute(RestRequest request, ResponseHandler<T> successResponseHandler) {
    HttpURLConnection connection;
    try {
      connection = prepareConnection(request);
      if (request.hasOutput()) {
        request.getOutput().prepareConnection(connection);
      }
    } catch (IOException e) {
      return new ErrorResponse<>();
    }

    Map<String, List<String>> responseHeaders = new HashMap<>();
    int responseCode = RestResponse.REQUEST_NOT_SENT;
    try {

      responseCode = connection.getResponseCode();
      responseHeaders.putAll(connection.getHeaderFields());
      boolean isErrorCode = IOUtils.isErrorCode(responseCode);
      try (InputStream is = isErrorCode ?
          connection.getErrorStream() :
          connection.getInputStream()) {
        if (isErrorCode) {
          return  new ErrorResponse<>(responseCode, new HttpHeaders(responseHeaders), IOUtils.toBytes(is));
        } else {
          return new SuccessResponse<>(responseCode, new HttpHeaders(responseHeaders), successResponseHandler.convert(is));
        }
      }
    } catch (IOException e) {
      return new ErrorResponse<>(responseCode, new HttpHeaders(responseHeaders));
    } finally {
      connection.disconnect();
    }
  }

  public void setJwt(String jwt) {
    this.jwt = jwt;
  }

  public void removeJwt() {
    setJwt(null);
  }

  public String getJwt() {
    return jwt;
  }


  private static abstract class AbstractRestResponse<T> implements RestResponse<T> {

    private final int responseCode;
    private final HttpHeaders headers;

    public AbstractRestResponse(int responseCode, HttpHeaders headers) {
      this.responseCode = responseCode;
      this.headers = headers;
    }

    @Override
    public HttpHeaders getHeaders() {
      return headers;
    }

    @Override
    public int getResponseCode() {
      return responseCode;
    }

  }

  private static class SuccessResponse<T> extends AbstractRestResponse<T> {

    private final T data;

    SuccessResponse(int responseCode, HttpHeaders headers, T data) {
      super(responseCode, headers);
      this.data = data;
    }

    @Override
    public boolean isErrorResponse() {
      return false;
    }

    @Override
    public boolean isSuccessful() {
      return true;
    }

    @Override
    public T getData() {
      return data;
    }

    @Override
    public <ErrorT> ErrorT getErrorData(ResponseHandler<ErrorT> responseHandler) {
      return null;
    }

    @Override
    public byte[] getRawErrorData() {
      return null;
    }
  }

  private class ErrorResponse<T> extends AbstractRestResponse<T> {

    private final byte[] bytes;

    ErrorResponse() { //no response
      this(REQUEST_NOT_SENT, new HttpHeaders(Collections.<String, List<String>>emptyMap()));
    }

    ErrorResponse(int responseCode, HttpHeaders headers) {
      this(responseCode, headers, null);
    }

    ErrorResponse(int responseCode, HttpHeaders headers, byte[] bytes) {
      super(responseCode, headers);
      this.bytes = bytes;
    }

    @Override
    public boolean isErrorResponse() {
      return true;
    }

    @Override
    public boolean isSuccessful() {
      return false;
    }

    @Override
    public T getData() {
      return null;
    }

    @Override
    public <ErrorT> ErrorT getErrorData(ResponseHandler<ErrorT> responseHandler) {
      if (bytes == null) {
        return null;
      }
      try (ByteArrayInputStream is = new ByteArrayInputStream(bytes)) {
        return responseHandler.convert(is);
      } catch (IOException e) {
        return null;
      }
    }

    @Override
    public byte[] getRawErrorData() {
      return bytes;
    }

  }
}

File: src/main/kotlin/com/tambapps/http/client/AbstractRestClient.kt
package com.tambapps.http.client

import com.tambapps.http.client.request.Request
import com.tambapps.http.client.response.ErrorResponse
import com.tambapps.http.client.response.HttpHeaders
import com.tambapps.http.client.response.Response
import com.tambapps.http.client.response.SuccessResponse
import com.tambapps.http.client.response.handler.ResponseHandler
import com.tambapps.http.client.util.IOUtils
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

abstract class AbstractRestClient(baseUrl: String) {

    // ends without '/'
    private val baseUrl: String =
            if (baseUrl.endsWith("/")) baseUrl.substring(0, baseUrl.length - 1)
            else baseUrl
    var jwt: String? = null


    @Throws(MalformedURLException::class)
    private fun getUrl(endpoint: String): URL {
        return URL(
                if (endpoint.startsWith("/")) baseUrl + endpoint
                else "$baseUrl/$endpoint"
        )
    }

    @Throws(IOException::class)
    private fun prepareConnection(request: Request): HttpURLConnection {
        val connection = getUrl(request.endpoint).openConnection() as HttpURLConnection
        connection.requestMethod = request.method
        for ((key, value) in request.headers) {
            connection.setRequestProperty(key, value)
        }
        if (jwt != null) {
            connection.setRequestProperty("Authorization", "Bearer $jwt")
        }
        if (request.timeout != null) {
            connection.connectTimeout = request.timeout
        }
        return connection
    }

    protected open fun <T> doExecute(request: Request, successResponseHandler: ResponseHandler<T>): Response<T> {
        val connection: HttpURLConnection
        try {
            connection = prepareConnection(request)
            if (request.hasBody()) {
                request.bodyProcessor!!.prepareConnection(connection)
            }
        } catch (e: IOException) {
            return ErrorResponse(e.message!!.toByteArray())
        }
        val responseHeaders: MutableMap<String, List<String>> = HashMap()
        var responseCode = Response.REQUEST_NOT_SENT
        try {
            responseCode = connection.responseCode
            responseHeaders.putAll(connection.headerFields)
            val isErrorCode = IOUtils.isErrorCode(responseCode)
            if (isErrorCode) {
                val stream = connection.errorStream
                if (stream != null) {
                    return stream.use {
                        ErrorResponse(responseCode, HttpHeaders(responseHeaders), IOUtils.toBytes(it))
                    }
                } else {
                    return ErrorResponse(responseCode, HttpHeaders(responseHeaders), ByteArray(0))
                }
            } else {
                val stream = connection.inputStream
                if (stream != null) {
                    return SuccessResponse(responseCode, HttpHeaders(responseHeaders), successResponseHandler.convert(stream))
                } else {
                    return SuccessResponse(responseCode, HttpHeaders(responseHeaders), null)
                }
            }
        } catch (e: IOException) {
            return ErrorResponse(responseCode, HttpHeaders(responseHeaders), e.message!!.toByteArray())
        } finally {
            connection.disconnect()
        }
    }

    fun removeJwt() {
        this.jwt = null
    }
}